src_lockfree/htbl/htbl_unsafe.ml

```diff
@@ -0,0 +1,634 @@
+
+(** This must be called with [r.pending == Nothing]. *)
+let[@inline never] try_resize t r new_capacity ~clear =
+  (* We must make sure that on every resize we use a physically different
+     [Resize _] value to indicate unprocessed target buckets.  The use of
+     [Sys.opaque_identity] below ensures that a new value is allocated. *)
+  let resize_avoid_aba =
+    if clear then B (Nil_with_size { size_modifier = Size.used_once })
+    else
+      B
+        (Resize
+           {
+             spine =
+               Sys.opaque_identity
+                 (Nil_with_size { size_modifier = Size.used_once });
+           })
+  in
+  let buckets = Atomic_array.make new_capacity resize_avoid_aba in
+  let new_r =
+    {
+      r with
+      pending =
+        Resize { buckets; size = (if clear then Size.create () else r.size) };
+    }
+  in
+  Atomic.compare_and_set t r new_r
+  && begin
+       finish t new_r |> ignore;
+       true
+     end
+
+let adjust_size t r mask result =
+  if
+    r.pending == Nothing
+    && Int64.to_int (Random.bits64 ()) land mask = 0
+    && Atomic.get t == r
+  then begin
+    (* Reading the size is potentially expensive, so we only check it
+       occasionally.  The bigger the table the less frequently we should need to
+       resize. *)
+    let size = Size.get r.size in
+    let capacity = Atomic_array.length r.buckets in
+    if capacity < size && capacity < r.max_buckets then
+      try_resize t r (capacity + capacity) ~clear:false |> ignore
+    else if r.min_buckets < capacity && size + size + size < capacity then
+      try_resize t r (capacity lsr 1) ~clear:false |> ignore
+  end;
+  result
+
+(* *)
+
+(** [get] only returns with a state where [pending = Nothing]. *)
+let[@inline] get t =
+  let r = Atomic.get t in
+  if r.pending == Nothing then r else finish t r
+
+(* *)
+
+let rec assoc_node t key = function
+  | Nil -> (Nil : (_, _, [< `Nil | `Cons ]) tdt)
+  | Cons r as cons -> if t r.key key then cons else assoc_node t key r.rest
+
+let find_node t key =
+  (* Reads can proceed in parallel with writes. *)
+  let r = Atomic.get t in
+  let h = r.hash key in
+  let mask = Atomic_array.length r.buckets - 1 in
+  let i = h land mask in
+  let rec loop = function
+    | B (Nil_with_size nil_r) ->
+        if nil_r.size_modifier != Size.used_once then
+          Size.update_once r.size nil_r.size_modifier;
+        Nil
+    | B (Cons_with_size cons_r) ->
+        if cons_r.size_modifier != Size.used_once then
+          Size.update_once r.size cons_r.size_modifier;
+        if r.equal cons_r.key key then
+          Cons { key = cons_r.key; value = cons_r.value; rest = cons_r.rest }
+        else assoc_node r.equal key cons_r.rest
+    | B (Resize resize_r) ->
+        (* A resize is in progress.  The spine of the resize still holds what was
+           in the bucket before resize reached that bucket. *)
+        loop (B resize_r.spine)
+  in
+
+  loop @@ Atomic_array.unsafe_fenceless_get r.buckets i
```

Comment:
Having an inner function like this with the native compiler generates a heap allocation for a closure.

I have this in `dune-workspace` at the root of project:

```
(lang dune 3.9)

(env
 (dev
  (ocamlopt_flags
   (:standard -S))
  (flags
   (:standard -warn-error -A)))
 (release
  (ocamlopt_flags
   (:standard -S))))
```

This instructs the compiler to produce assembly output (`*.s` files) that you can then find under the `_build` directory hierarchy.

Looking at the generated code (this is actually from the safe version), I can see the following (I've trimmed the output a bit):

```
_camlSaturn_lockfree__Htbl.find_node_926:
;; [SNIP]
	ldr	x16, [x28, #0]  ;; This starts the allocation
	sub	x27, x27, #40
	cmp	x27, x16
	b.lo	L404                ;; This jumps to do GC if necessary (see the last two lines)
L403:	add	x12, x27, #8
	str	x12, [sp, #0]
	movz	x13, #4343, lsl #0
	str	x13, [x12, #-8]
	adrp	x14, _camlSaturn_lockfree__Htbl.loop_934@GOTPAGE
	ldr	x14, [x14, _camlSaturn_lockfree__Htbl.loop_934@GOTPAGEOFF]
	str	x14, [x12, #0]
	movz	x15, #5, lsl #0
	movk	x15, #256, lsl #48
	str	x15, [x12, #8]
	ldr	x22, [sp, #8]
	str	x22, [x12, #16]
	str	x21, [x12, #24]
	mov	x0, x6
	.loc	1	409
	bl	_camlSaturn_lockfree__Htbl.unsafe_fenceless_get_376
L405:
	ldr	x1, [sp, #0]
	.loc	1	409
	ldr	x30, [sp, #24]
	add	sp, sp, #32
	.cfi_adjust_cfa_offset	-32
	b	_camlSaturn_lockfree__Htbl.loop_934
	.cfi_adjust_cfa_offset	32
L404:	bl	_caml_call_gc
L402:	b	L403
```

Doing allocations for `find_node` is unnecessary and is bad for performance.

I would recommend moving the loop to a top level function ([like in the original](https://github.com/ocaml-multicore/picos/blob/849dea9f61319265d947ab4c5c9d134cf3116a74/lib/picos_aux.htbl/picos_aux_htbl.ml#L346-L364)).

Reply:
Thanks ! It is actually an extra loop (compared to the original), that is needed because of the new type definition : `resize` can not call `assoc_node`. But I did not remember about this extra allocation!